### DISCUSSION TITLE
How to convert `at:://` URI to `https://` URL?

### DISCUSSION THREAD

**Discussion by Geobert** [bluesky-social/atproto#2523] [How to do it?] [answered]
Hi,

After creating a record, we get a `uri` with `at://` protocol but I want the URL to the created post. How do we get it?

**Comment by bnewbold** [comment, marked as answer]
I'll give a pragmatic answer below, but first a protocol philosophy framing question: does it even make sense for there to be a "canonical" web (HTTPS) URL for record AT URIs? atproto URIs have the "authority" section scoped to an individual account (DID or handle), intentionally not scoped to a service (eg, web server), because the hosting location for the account could change over time. We don't want to over-centralize and have any one server or service be the "official" location for content in the network.

One way to relax this would be to say that individual accounts do have a canonical host at any time, such as the handle or PDS host, and we could resolve AT-URIs to one of those locations to be redirected. That is nicely principled, but requires extra work by the account, and application-aware routing or display logic.

Another mechanism would be to assign relatively canonical service hostnames for specific NSIDs (aka, record schemas or collections). We could say that https://bsky.app is the pseudo-canonical host service for the `app.bsky.*` lexicons. That kind of makes sense, but is also kind of centralizing.

Pragmatically, today, if you have a bluesky post record and want to view it on the web, you probably want to link to https://bsky.app: it is the pseudo-canonical host for that content.

The process is usually to look at the AT URI authority (DID), collection, and rkey segments:

```
at://<DID>/<COLLECTION>/<RKEY>
```

Match on the collection: if it is `app.bsky.feed.post`, it is a post record. Create post URLs on bsky.app like:

```
https://bsky.app/profile/<DID>/post/<RKEY>
```

**Reply by mohamedmhiri** [reply]
thank you

**Reply by tcjr** [reply]
Thanks!  Here's the above in TypeScript:

```typescript
/**
 * Converts an AT URI for a Bluesky post to a https://bsky.app.
 *
 * @param atUri The AT URI of the post.  Must be in the format at://<DID>/<COLLECTION>/<RKEY>
 * @returns The HTTPS URL to view the post on bsky.app, or null if the AT URI is invalid or not a post.
 */
function atUriToBskyAppUrl(atUri: string): string | null {
  const regex = /^at:\/\/([^/]+)\/([^/]+)\/([^/]+)$/;
  const match = atUri.match(regex);

  if (!match) {
    return null; // Invalid AT URI format
  }

  const did = match[1];
  const collection = match[2];
  const rkey = match[3];

  if (collection === 'app.bsky.feed.post') {
    return `https://bsky.app/profile/${did}/post/${rkey}`;
  } else {
    return null; // Not a post record
  }
}
```

Usage:
```typescript
const validAtUri = "at://did:plc:6aapcgkhjeffsdjc656mshnp/app.bsky.feed.post/3lhlw4gq4uj2t";
const bskyAppUrl = atUriToBskyAppUrl(validAtUri);
if (bskyAppUrl) {
  console.log(`BSky App URL: ${bskyAppUrl}`);
} else {
  console.log("Invalid AT URI or not a post.");
}
```

**Reply by DavidBuchanan314** [reply]
by the way, here's more generic logic that can handle some non-bsky apps, too: https://github.com/notjuliet/pdsls/blob/74d45a3a56149d706fe950e2a7123a526d4ac5cf/src/views/record.tsx#L146-L197

**Comment by Geobert** [comment]
Thanks for this deep and complete answer!
